// @process-engine/management_api_http/dist/amd/endpoints/token_history/token_history_router.js
define(["require", "exports", "@essential-projects/http_node", "@process-engine/management_api_contracts", "async-middleware", "../../middlewares/resolve_identity"], function (require, exports, http_node_1, management_api_contracts_1, async_middleware_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class TokenHistoryRouter extends http_node_1.BaseRouter {
        constructor(tokenHistoryController, identityService) {
            super();
            this.tokenHistoryController = tokenHistoryController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.tokenHistoryController;
            this.router.get(management_api_contracts_1.restSettings.paths.getTokensForFlowNode, async_middleware_1.wrap(controller.getTokensForFlowNode.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getTokensForFlowNodeByProcessInstanceId, async_middleware_1.wrap(controller.getTokensForFlowNodeByProcessInstanceId.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getTokensForProcessInstance, async_middleware_1.wrap(controller.getTokensForProcessInstance.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getTokensForCorrelationAndProcessModel, async_middleware_1.wrap(controller.getTokensForCorrelationAndProcessModel.bind(controller)));
        }
    }
    exports.TokenHistoryRouter = TokenHistoryRouter;
});
