// @process-engine/consumer_api_contracts/dist/amd/messages/event_aggregator_settings.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /* eslint-disable max-len */
    exports.messageParams = {
        correlationId: ':correlation_id',
        endEventId: ':end_event_id',
        flowNodeInstanceId: ':flow_node_instance_id',
        userTaskId: ':user_task_id',
        manualTaskId: ':manual_task_id',
        messageReference: ':message_ref',
        processInstanceId: ':process_instance_id',
        processModelId: ':process_model_id',
        signalReference: ':signal_ref',
        emptyActivityId: ':empty_activity_id',
    };
    exports.messagePaths = {
        // Generic messages
        boundaryEventTriggered: 'boundary_event_triggered',
        intermediateThrowEventTriggered: 'intermediate_throw_event_triggered',
        intermediateCatchEventReached: 'intermediate_catch_event_reached',
        intermediateCatchEventFinished: 'intermediate_catch_event_finished',
        activityReached: 'activity_reached',
        activityFinished: 'activity_finished',
        emptyActivityReached: 'empty_activity_reached',
        emptyActivityFinished: 'empty_activity_finished',
        userTaskReached: 'user_task_reached',
        userTaskFinished: 'user_task_finished',
        manualTaskReached: 'manual_task_reached',
        manualTaskFinished: 'manual_task_finished',
        processStarted: 'process_started',
        processEnded: 'process_ended',
        processError: 'process_error',
        processTerminated: 'process_terminated',
        // instance specific messages
        finishEmptyActivity: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/emptyactivity/${exports.messageParams.flowNodeInstanceId}/finish`,
        emptyActivityWithInstanceIdFinished: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/emptyactivity/${exports.messageParams.flowNodeInstanceId}/finished`,
        finishUserTask: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/usertask/${exports.messageParams.flowNodeInstanceId}/finish`,
        userTaskWithInstanceIdFinished: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/usertask/${exports.messageParams.flowNodeInstanceId}/finished`,
        finishManualTask: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/manualtask/${exports.messageParams.flowNodeInstanceId}/finish`,
        manualTaskWithInstanceIdFinished: `/processengine/correlation/${exports.messageParams.correlationId}/processinstance/${exports.messageParams.processInstanceId}/manualtask/${exports.messageParams.flowNodeInstanceId}/finished`,
        endEventReached: `/processengine/correlation/${exports.messageParams.correlationId}/processmodel/${exports.messageParams.processModelId}/ended`,
        messageEventReached: `/processengine/process/message/${exports.messageParams.messageReference}`,
        signalEventReached: `/processengine/process/signal/${exports.messageParams.signalReference}`,
        terminateEndEventReached: `/processengine/process/${exports.messageParams.processInstanceId}/terminated`,
        processInstanceStarted: `/processengine/process_started/${exports.messageParams.processModelId}`,
        processInstanceEnded: `/processengine/process/${exports.messageParams.processInstanceId}/ended`,
    };
});
