// @process-engine/process_engine_contracts/dist/amd/runtime/messages/system_events/base_system_event_message.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * The base class for definining system event messages.
     * System events are peripheral events that are not contained within a
     * ProcessModel, but are relevant to its execution.
     *
     * For example: A UserTask was reached or finished.
     */
    class BaseSystemEventMessage {
        constructor(correlationId, processModelId, processInstanceId, flowNodeId, flowNodeInstanceId, processInstanceOwner, currentToken) {
            this.correlationId = correlationId;
            this.processModelId = processModelId;
            this.processInstanceId = processInstanceId;
            this.flowNodeId = flowNodeId;
            this.flowNodeInstanceId = flowNodeInstanceId;
            this.processInstanceOwner = processInstanceOwner;
            this.currentToken = currentToken;
        }
    }
    exports.BaseSystemEventMessage = BaseSystemEventMessage;
});
