// @process-engine/process_engine_core/dist/amd/model/parser/definitions_parser.js
define(["require", "exports", "@process-engine/persistence_api.contracts", "../type_factory", "./index"], function (require, exports, persistence_api_contracts_1, type_factory_1, index_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    function parseDefinitions(parsedObjectModel) {
        const definitions = parsedObjectModel[persistence_api_contracts_1.BpmnTags.CommonElement.Definitions];
        return convertToInternalObjectModel(definitions);
    }
    exports.parseDefinitions = parseDefinitions;
    function convertToInternalObjectModel(parsedXml) {
        const definitions = createDefinitionBaseObject(parsedXml);
        definitions.collaboration = index_1.parseCollaboration(parsedXml);
        definitions.processes = index_1.parseProcesses(parsedXml);
        return definitions;
    }
    function createDefinitionBaseObject(parsedXml) {
        const basicDefinition = type_factory_1.createObjectWithCommonProperties(parsedXml, persistence_api_contracts_1.Model.Definitions);
        basicDefinition.xmlns = {
            bpmn: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.bpmn],
            bpmndi: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.bpmndi],
            camunda: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.camunda],
            dc: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.dc],
            di: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.di],
            xsi: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.xsi],
        };
        basicDefinition.targetNamespace = parsedXml.targetNamespace;
        basicDefinition.exporter = parsedXml.exporter;
        basicDefinition.exporterVersion = parsedXml.exporterVersion;
        return basicDefinition;
    }
});
