// @process-engine/management_api_http/dist/commonjs/endpoints/empty_activity/empty_activity_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_middleware_1 = require("async-middleware");
const http_node_1 = require("@essential-projects/http_node");
const management_api_contracts_1 = require("@process-engine/management_api_contracts");
const resolve_identity_1 = require("../../middlewares/resolve_identity");
class EmptyActivityRouter extends http_node_1.BaseRouter {
    constructor(emptyActivityController, identityService) {
        super();
        this.emptyActivityController = emptyActivityController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/management/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        this.registerRoutes();
    }
    registerMiddlewares() {
        const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
    registerRoutes() {
        const controller = this.emptyActivityController;
        this.router.get(management_api_contracts_1.restSettings.paths.processModelEmptyActivities, async_middleware_1.wrap(controller.getEmptyActivitiesForProcessModel.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.processInstanceEmptyActivities, async_middleware_1.wrap(controller.getEmptyActivitiesForProcessInstance.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.correlationEmptyActivities, async_middleware_1.wrap(controller.getEmptyActivitiesForCorrelation.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.processModelCorrelationEmptyActivities, async_middleware_1.wrap(controller.getEmptyActivitiesForProcessModelInCorrelation.bind(controller)));
        this.router.post(management_api_contracts_1.restSettings.paths.finishEmptyActivity, async_middleware_1.wrap(controller.finishEmptyActivity.bind(controller)));
    }
}
exports.EmptyActivityRouter = EmptyActivityRouter;
