// @process-engine/logging_api_core/dist/commonjs/logging_api_service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class LoggingApiService {
    constructor(iamService, loggingRepository) {
        this.iamService = iamService;
        this.loggingRepository = loggingRepository;
    }
    // TODO: Add claim checks as soon as necessary claims have been defined.
    async readLogForProcessModel(identity, processModelId) {
        return this.loggingRepository.readLogForProcessModel(processModelId);
    }
    async writeLogForProcessModel(correlationId, processModelId, processInstanceId, logLevel, measuredAt, message, timestamp, error) {
        await this
            .loggingRepository
            .writeLogForProcessModel(correlationId, processModelId, processInstanceId, logLevel, measuredAt, message, timestamp, error);
    }
    async writeLogForFlowNode(correlationId, processModelId, processInstanceId, flowNodeInstanceId, flowNodeId, logLevel, measuredAt, tokenPayload, message, timestamp, error) {
        await this.loggingRepository.writeLogForFlowNode(correlationId, processModelId, processInstanceId, flowNodeInstanceId, flowNodeId, logLevel, measuredAt, tokenPayload, message, timestamp, error);
    }
    async archiveProcessModelLogs(identity, processModelId) {
        await this
            .loggingRepository
            .archiveProcessModelLogs(processModelId);
    }
}
exports.LoggingApiService = LoggingApiService;
