// @process-engine/management_api_http/dist/commonjs/endpoints/cronjob/cronjob_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_middleware_1 = require("async-middleware");
const http_node_1 = require("@essential-projects/http_node");
const management_api_contracts_1 = require("@process-engine/management_api_contracts");
const resolve_identity_1 = require("../../middlewares/resolve_identity");
class CronjobRouter extends http_node_1.BaseRouter {
    constructor(cronjobController, identityService) {
        super();
        this.cronjobController = cronjobController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/management/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        this.registerRoutes();
    }
    registerMiddlewares() {
        const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
    registerRoutes() {
        const controller = this.cronjobController;
        this.router.get(management_api_contracts_1.restSettings.paths.getActiveCronjobs, async_middleware_1.wrap(controller.getAllActiveCronjobs.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.getCronjobExecutionHistoryForProcessModel, async_middleware_1.wrap(controller.getCronjobExecutionHistoryForProcessModel.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.getCronjobExecutionHistoryForCrontab, async_middleware_1.wrap(controller.getCronjobExecutionHistoryForCrontab.bind(controller)));
    }
}
exports.CronjobRouter = CronjobRouter;
