// @process-engine/management_api_http/dist/amd/endpoints/correlation/correlation_router.js
define(["require", "exports", "async-middleware", "@essential-projects/http_node", "@process-engine/management_api_contracts", "../../middlewares/resolve_identity"], function (require, exports, async_middleware_1, http_node_1, management_api_contracts_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class CorrelationRouter extends http_node_1.BaseRouter {
        constructor(correlationController, identityService) {
            super();
            this.correlationController = correlationController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.correlationController;
            this.router.get(management_api_contracts_1.restSettings.paths.getAllCorrelations, async_middleware_1.wrap(controller.getAllCorrelations.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveCorrelations, async_middleware_1.wrap(controller.getActiveCorrelations.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getCorrelationById, async_middleware_1.wrap(controller.getCorrelationById.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getCorrelationsByProcessModelId, async_middleware_1.wrap(controller.getCorrelationsByProcessModelId.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getProcessInstanceById, async_middleware_1.wrap(controller.getProcessInstanceById.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getProcessInstancesForCorrelation, async_middleware_1.wrap(controller.getProcessInstancesForCorrelation.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getProcessInstancesForProcessModel, async_middleware_1.wrap(controller.getProcessInstancesForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getProcessInstancesByState, async_middleware_1.wrap(controller.getProcessInstancesByState.bind(controller)));
        }
    }
    exports.CorrelationRouter = CorrelationRouter;
});
