// @process-engine/persistence_api.contracts/dist/commonjs/data_models/process_model/activities/service_task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const activity_1 = require("./activity");
/**
 * Describes a BPMN ServiceTask.
 * A ServiceTask is used to call either a function from a specific class,
 * or for performing web requests.
 */
class ServiceTask extends activity_1.Activity {
    get bpmnType() {
        return constants_1.BpmnType.serviceTask;
    }
}
exports.ServiceTask = ServiceTask;
/**
 * Describes all currently supported ServiceTask types.
 */
var ServiceTaskType;
(function (ServiceTaskType) {
    ServiceTaskType["external"] = "external";
    ServiceTaskType["internal"] = "internal";
})(ServiceTaskType = exports.ServiceTaskType || (exports.ServiceTaskType = {}));
/**
 * The base class for all invocation types.
 */
// eslint-disable-next-line @typescript-eslint/no-extraneous-class
class Invocation {
}
exports.Invocation = Invocation;
/**
 * Used by ServiceTasks.
 * This contains the definition for a method invocation.
 */
class MethodInvocation extends Invocation {
}
exports.MethodInvocation = MethodInvocation;
