// @process-engine/management_api_http/dist/amd/endpoints/kpi/kpi_router.js
define(["require", "exports", "@essential-projects/http_node", "@process-engine/management_api_contracts", "async-middleware", "../../middlewares/resolve_identity"], function (require, exports, http_node_1, management_api_contracts_1, async_middleware_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class KpiRouter extends http_node_1.BaseRouter {
        constructor(kpiController, identityService) {
            super();
            this.kpiController = kpiController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.kpiController;
            this.router.get(management_api_contracts_1.restSettings.paths.getRuntimeInformationForProcessModel, async_middleware_1.wrap(controller.getRuntimeInformationForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveTokensForProcessModel, async_middleware_1.wrap(controller.getActiveTokensForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getRuntimeInformationForFlowNode, async_middleware_1.wrap(controller.getRuntimeInformationForFlowNode.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveTokensForFlowNode, async_middleware_1.wrap(controller.getActiveTokensForFlowNode.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveTokensForProcessInstance, async_middleware_1.wrap(controller.getActiveTokensForProcessInstance.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveTokensForCorrelationAndProcessModel, async_middleware_1.wrap(controller.getActiveTokensForCorrelationAndProcessModel.bind(controller)));
        }
    }
    exports.KpiRouter = KpiRouter;
});
