// @process-engine/consumer_api_http/dist/commonjs/endpoints/external_task/external_task_router_deprecated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const async_middleware_1 = require("async-middleware");
const http_node_1 = require("@essential-projects/http_node");
const consumer_api_contracts_1 = require("@process-engine/consumer_api_contracts");
const index_1 = require("../../middlewares/index");
/**
 * This router provides the old endpoints for the ExternalTaskAPI.
 * Those routes will be removed in future versions, after users had time
 * to adjust their usage of ExternalTasks.
 */
class ExternalTaskRouterDeprecated extends http_node_1.BaseRouter {
    constructor(externalTaskController, identityService) {
        super();
        this.externalTaskController = externalTaskController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/external_task/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        const controller = this.externalTaskController;
        this.router.post(consumer_api_contracts_1.restSettings.paths.fetchAndLockExternalTasks, async_middleware_1.wrap(controller.fetchAndLockExternalTasks.bind(controller)));
        this.router.post(consumer_api_contracts_1.restSettings.paths.extendExternalTaskLock, async_middleware_1.wrap(controller.extendLock.bind(controller)));
        this.router.post(consumer_api_contracts_1.restSettings.paths.finishExternalTaskWithBpmnError, async_middleware_1.wrap(controller.handleBpmnError.bind(controller)));
        this.router.post(consumer_api_contracts_1.restSettings.paths.finishExternalTaskWithServiceError, async_middleware_1.wrap(controller.handleServiceError.bind(controller)));
        this.router.post(consumer_api_contracts_1.restSettings.paths.finishExternalTask, async_middleware_1.wrap(controller.finishExternalTask.bind(controller)));
    }
    registerMiddlewares() {
        const resolveIdentity = index_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
}
exports.ExternalTaskRouterDeprecated = ExternalTaskRouterDeprecated;
