// @process-engine/management_api_http/dist/amd/endpoints/process_model/process_model_router.js
define(["require", "exports", "@essential-projects/http_node", "@process-engine/management_api_contracts", "async-middleware", "../../middlewares/resolve_identity"], function (require, exports, http_node_1, management_api_contracts_1, async_middleware_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class ProcessModelRouter extends http_node_1.BaseRouter {
        constructor(processModelController, identityService) {
            super();
            this.processModelController = processModelController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.processModelController;
            this.router.get(management_api_contracts_1.restSettings.paths.processModels, async_middleware_1.wrap(controller.getProcessModels.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.processModelById, async_middleware_1.wrap(controller.getProcessModelById.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.processModelByProcessInstanceId, async_middleware_1.wrap(controller.getProcessModelByProcessInstanceId.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.processModelStartEvents, async_middleware_1.wrap(controller.getStartEventsForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.deleteProcessDefinitionsByProcessModelId, async_middleware_1.wrap(controller.deleteProcessDefinitionsByProcessModelId.bind(controller)));
            this.router.post(management_api_contracts_1.restSettings.paths.startProcessInstance, async_middleware_1.wrap(controller.startProcessInstance.bind(controller)));
            this.router.post(management_api_contracts_1.restSettings.paths.updateProcessDefinitionsByName, async_middleware_1.wrap(controller.updateProcessDefinitionsByName.bind(controller)));
            this.router.post(management_api_contracts_1.restSettings.paths.terminateProcessInstance, async_middleware_1.wrap(controller.terminateProcessInstance.bind(controller)));
        }
    }
    exports.ProcessModelRouter = ProcessModelRouter;
});
