// @process-engine/management_api_http/dist/amd/endpoints/process_model/process_model_controller.js
define(["require", "exports", "@process-engine/management_api_contracts"], function (require, exports, management_api_contracts_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class ProcessModelController {
        constructor(processModelService) {
            this.httpCodeSuccessfulResponse = 200;
            this.httpCodeSuccessfulNoContentResponse = 204;
            this.processModelService = processModelService;
        }
        async getProcessModels(request, response) {
            const identity = request.identity;
            const offset = request.query.offset || 0;
            const limit = request.query.limit || 0;
            const result = await this.processModelService.getProcessModels(identity, offset, limit);
            response.status(this.httpCodeSuccessfulResponse).json(result);
        }
        async getProcessModelById(request, response) {
            const processModelId = request.params.process_model_id;
            const identity = request.identity;
            const result = await this.processModelService.getProcessModelById(identity, processModelId);
            response.status(this.httpCodeSuccessfulResponse).json(result);
        }
        async getProcessModelByProcessInstanceId(request, response) {
            const processInstanceId = request.params.process_instance_id;
            const identity = request.identity;
            const result = await this.processModelService.getProcessModelByProcessInstanceId(identity, processInstanceId);
            response.status(this.httpCodeSuccessfulResponse).json(result);
        }
        async getStartEventsForProcessModel(request, response) {
            const processModelId = request.params.process_model_id;
            const identity = request.identity;
            const result = await this.processModelService.getStartEventsForProcessModel(identity, processModelId);
            response.status(this.httpCodeSuccessfulResponse).json(result);
        }
        async startProcessInstance(request, response) {
            const processModelId = request.params.process_model_id;
            const startEventId = request.query.start_event_id;
            const endEventId = request.query.end_event_id;
            const payload = request.body;
            let startCallbackType = Number.parseInt(request.query.start_callback_type);
            if (!startCallbackType) {
                startCallbackType = management_api_contracts_1.DataModels.ProcessModels.StartCallbackType.CallbackOnProcessInstanceCreated;
            }
            const identity = request.identity;
            const result = await this.processModelService.startProcessInstance(identity, processModelId, payload, startCallbackType, startEventId, endEventId);
            response.status(this.httpCodeSuccessfulResponse).json(result);
        }
        async updateProcessDefinitionsByName(request, response) {
            const processDefinitionsName = request.params.process_definitions_name;
            const payload = request.body;
            const identity = request.identity;
            await this.processModelService.updateProcessDefinitionsByName(identity, processDefinitionsName, payload);
            response.status(this.httpCodeSuccessfulNoContentResponse).send();
        }
        async deleteProcessDefinitionsByProcessModelId(request, response) {
            const processDefinitionsName = request.params.process_model_id;
            const identity = request.identity;
            await this.processModelService.deleteProcessDefinitionsByProcessModelId(identity, processDefinitionsName);
            response.status(this.httpCodeSuccessfulNoContentResponse).send();
        }
        async terminateProcessInstance(request, response) {
            const processInstanceId = request.params.process_instance_id;
            const identity = request.identity;
            await this.processModelService.terminateProcessInstance(identity, processInstanceId);
            response.status(this.httpCodeSuccessfulNoContentResponse).send();
        }
    }
    exports.ProcessModelController = ProcessModelController;
});
