// @process-engine/process_engine_core/dist/commonjs/model/parser/definitions_parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const persistence_api_contracts_1 = require("@process-engine/persistence_api.contracts");
const type_factory_1 = require("../type_factory");
const index_1 = require("./index");
function parseDefinitions(parsedObjectModel) {
    const definitions = parsedObjectModel[persistence_api_contracts_1.BpmnTags.CommonElement.Definitions];
    return convertToInternalObjectModel(definitions);
}
exports.parseDefinitions = parseDefinitions;
function convertToInternalObjectModel(parsedXml) {
    const definitions = createDefinitionBaseObject(parsedXml);
    definitions.collaboration = index_1.parseCollaboration(parsedXml);
    definitions.processes = index_1.parseProcesses(parsedXml);
    return definitions;
}
function createDefinitionBaseObject(parsedXml) {
    const basicDefinition = type_factory_1.createObjectWithCommonProperties(parsedXml, persistence_api_contracts_1.Model.Definitions);
    basicDefinition.xmlns = {
        bpmn: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.bpmn],
        bpmndi: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.bpmndi],
        camunda: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.camunda],
        dc: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.dc],
        di: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.di],
        xsi: parsedXml[persistence_api_contracts_1.BpmnTags.XmlnsProperty.xsi],
    };
    basicDefinition.targetNamespace = parsedXml.targetNamespace;
    basicDefinition.exporter = parsedXml.exporter;
    basicDefinition.exporterVersion = parsedXml.exporterVersion;
    return basicDefinition;
}
