// @process-engine/consumer_api_http/dist/commonjs/endpoints/process_model/process_model_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const http_node_1 = require("@essential-projects/http_node");
const consumer_api_contracts_1 = require("@process-engine/consumer_api_contracts");
const async_middleware_1 = require("async-middleware");
const resolve_identity_1 = require("../../middlewares/resolve_identity");
class ProcessModelRouter extends http_node_1.BaseRouter {
    constructor(processModelController, identityService) {
        super();
        this.processModelController = processModelController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/consumer/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        this.registerRoutes();
    }
    registerMiddlewares() {
        const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
    registerRoutes() {
        const controller = this.processModelController;
        this.router.get(consumer_api_contracts_1.restSettings.paths.processModels, async_middleware_1.wrap(controller.getProcessModels.bind(controller)));
        this.router.get(consumer_api_contracts_1.restSettings.paths.processModelById, async_middleware_1.wrap(controller.getProcessModelById.bind(controller)));
        this.router.get(consumer_api_contracts_1.restSettings.paths.processModelByProcessInstanceId, async_middleware_1.wrap(controller.getProcessModelByProcessInstanceId.bind(controller)));
        this.router.get(consumer_api_contracts_1.restSettings.paths.getProcessResultForCorrelation, async_middleware_1.wrap(controller.getProcessResultForCorrelation.bind(controller)));
        this.router.get(consumer_api_contracts_1.restSettings.paths.getOwnProcessInstances, async_middleware_1.wrap(controller.getProcessInstancesByIdentity.bind(controller)));
        this.router.post(consumer_api_contracts_1.restSettings.paths.startProcessInstance, async_middleware_1.wrap(controller.startProcessInstance.bind(controller)));
    }
}
exports.ProcessModelRouter = ProcessModelRouter;
