// @process-engine/management_api_http/dist/commonjs/endpoints/flow_node_instance/flow_node_instance_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const http_node_1 = require("@essential-projects/http_node");
const management_api_contracts_1 = require("@process-engine/management_api_contracts");
const async_middleware_1 = require("async-middleware");
const resolve_identity_1 = require("../../middlewares/resolve_identity");
class FlowNodeInstanceRouter extends http_node_1.BaseRouter {
    constructor(flowNodeInstanceController, identityService) {
        super();
        this.flowNodeInstanceController = flowNodeInstanceController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/management/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        this.registerRoutes();
    }
    registerMiddlewares() {
        const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
    registerRoutes() {
        const controller = this.flowNodeInstanceController;
        this.router.get(management_api_contracts_1.restSettings.paths.getFlowNodeInstancesForProcessInstance, async_middleware_1.wrap(controller.getFlowNodeInstancesForProcessInstance.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.allSuspendedTasks, async_middleware_1.wrap(controller.getAllSuspendedTasks.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.suspendedProcessModelTasks, async_middleware_1.wrap(controller.getSuspendedTasksForProcessModel.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.suspendedProcessInstanceTasks, async_middleware_1.wrap(controller.getSuspendedTasksForProcessInstance.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.suspendedCorrelationTasks, async_middleware_1.wrap(controller.getSuspendedTasksForCorrelation.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.suspendedProcessModelCorrelationTasks, async_middleware_1.wrap(controller.getSuspendedTasksForProcessModelInCorrelation.bind(controller)));
    }
}
exports.FlowNodeInstanceRouter = FlowNodeInstanceRouter;
