// addict-ioc/dist/amd/registry.js
define(["require", "exports", "merge", "./default_settings", "./registration", "./registration_context", "./registration_settings"], function (require, exports, merge, default_settings_1, registration_1, registration_context_1, registration_settings_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var expectedRegistrationType;
    (function (expectedRegistrationType) {
        expectedRegistrationType["class"] = "class";
        expectedRegistrationType["factory"] = "factory";
        expectedRegistrationType["object"] = "object";
    })(expectedRegistrationType || (expectedRegistrationType = {}));
    var Registry = (function () {
        function Registry(settings, parentRegistry) {
            this.parentRegistry = parentRegistry;
            this.registrations = {};
            this.settings = settings;
            this.parentRegistry = parentRegistry;
        }
        Registry.prototype.initialize = function () {
            this.settings = this._mergeSettings(default_settings_1.defaultSettings, this.settings);
        };
        Registry.prototype.clear = function () {
            this.registrations = {};
        };
        Registry.prototype._mergeSettings = function (existingSettings, newSettings) {
            if (!existingSettings) {
                return newSettings;
            }
            if (!newSettings) {
                return existingSettings;
            }
            var mergedSettings = merge(true, existingSettings);
            Object.assign(mergedSettings, newSettings);
            Object.assign(mergedSettings.defaults, existingSettings.defaults);
            Object.assign(mergedSettings.defaults, newSettings.defaults);
            return mergedSettings;
        };
        Registry.prototype.importRegistrations = function (registrationSettings) {
            for (var _i = 0, registrationSettings_1 = registrationSettings; _i < registrationSettings_1.length; _i++) {
                var registrationSetting = registrationSettings_1[_i];
                var registration = new registration_1.Registration(registrationSetting);
                this.cacheRegistration(registrationSetting.key, registration);
            }
        };
        Registry.prototype.exportRegistrations = function (keysToExport) {
            var _this = this;
            var registrationKeys = keysToExport || this.getRegistrationKeys();
            return registrationKeys.map(function (registrationKey) {
                var registration = _this.getRegistration(registrationKey);
                var exportedSettings = merge(true, registration.settings);
                delete exportedSettings.type;
                delete exportedSettings.object;
                delete exportedSettings.factory;
                delete exportedSettings.resolver;
                return exportedSettings;
            });
        };
        Registry.prototype.isRegistered = function (key) {
            var registration = this.getRegistration(key);
            return !!registration;
        };
        Registry.prototype.createRegistrationTemplate = function (registrationSettings) {
            return new registration_context_1.RegistrationContext(this, registrationSettings);
        };
        Registry.prototype.register = function (key, type, settings) {
            this._validateRegistration(key, type, expectedRegistrationType.class);
            var registration = this.createRegistration(key, type, settings);
            this.cacheRegistration(key, registration);
            return registration;
        };
        Registry.prototype.registerObject = function (key, object, settings) {
            this._validateRegistration(key, object, expectedRegistrationType.object);
            var registration = this.createObjectRegistration(key, object, settings);
            this.cacheRegistration(key, registration);
            return registration;
        };
        Registry.prototype.registerFactory = function (key, factoryMethod, settings) {
            this._validateRegistration(key, factoryMethod, expectedRegistrationType.factory);
            var registration = this.createFactoryRegistration(key, factoryMethod, settings);
            this.cacheRegistration(key, registration);
            return registration;
        };
        Registry.prototype.unregister = function (key) {
            var registration = this.getRegistration(key);
            this.deleteRegistration(key);
            return registration;
        };
        Registry.prototype.createRegistration = function (key, type, registrationSettings) {
            var settings = registrationSettings
                ? new registration_settings_1.TypeRegistrationSettings(merge(true, registrationSettings))
                : merge(true, this.settings.defaults);
            settings.key = key;
            settings.type = type;
            var registration = new registration_1.Registration(settings);
            return registration;
        };
        Registry.prototype.createObjectRegistration = function (key, object, registrationSettings) {
            var settings = registrationSettings
                ? new registration_settings_1.ObjectRegistrationSettings(registrationSettings)
                : merge(true, this.settings.defaults);
            settings.key = key;
            settings.isObject = true;
            settings.object = object;
            var registration = new registration_1.Registration(settings);
            return registration;
        };
        Registry.prototype.createFactoryRegistration = function (key, factoryFunction, registrationSettings) {
            var settings = registrationSettings
                ? new registration_settings_1.FactoryRegistrationSettings(registrationSettings)
                : merge(true, this.settings.defaults);
            settings.key = key;
            settings.isFactory = true;
            settings.factory = factoryFunction;
            var registration = new registration_1.Registration(settings);
            return registration;
        };
        Registry.prototype.getRegistration = function (key) {
            var registration = this.registrations[key];
            if (!registration && this.parentRegistry) {
                return this.parentRegistry.getRegistration(key);
            }
            return registration;
        };
        Registry.prototype.getRegistrationKeys = function () {
            var keys = Object.keys(this.registrations);
            return this.sortKeys(keys);
        };
        Registry.prototype.sortKeys = function (keys) {
            return keys.sort(function (a, b) {
                if (a < b) {
                    return -1;
                }
                if (a > b) {
                    return 1;
                }
                return 0;
            });
        };
        Registry.prototype.cacheRegistration = function (key, registration) {
            this.registrations[key] = registration;
        };
        Registry.prototype.deleteRegistration = function (key) {
            delete this.registrations[key];
        };
        Registry.prototype.getKeysByTags = function () {
            var _this = this;
            var tags = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                tags[_i] = arguments[_i];
            }
            var registrationKeys = this.getRegistrationKeys();
            var foundKeys = [];
            var tagDictionary = this._buildTagDictionary.apply(this, tags);
            foundKeys = registrationKeys.filter(function (key) {
                var matchingRegistration = _this.getRegistration(key);
                var tagKeys = Object.keys(tagDictionary);
                var matchingTags = tagKeys.filter(function (tagKey) {
                    var registrationDoesNotHaveKey = !matchingRegistration.settings.tags[tagKey];
                    if (registrationDoesNotHaveKey) {
                        return false;
                    }
                    var tagValue = tagDictionary[tagKey];
                    var tagHasNoValue = tagValue.length < 1;
                    if (tagHasNoValue) {
                        return true;
                    }
                    var tagValuesMatch = tagValue === matchingRegistration.settings.tags[tagKey];
                    return tagValuesMatch;
                });
                var registrationHasAllMatchingTags = matchingTags.length === tagKeys.length;
                return registrationHasAllMatchingTags;
            });
            return this.sortKeys(foundKeys);
        };
        Registry.prototype._buildTagDictionary = function () {
            var tags = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                tags[_i] = arguments[_i];
            }
            var tagDictionary = {};
            for (var _a = 0, tags_1 = tags; _a < tags_1.length; _a++) {
                var tag = tags_1[_a];
                if (typeof tag === 'string') {
                    var hasTagDefaultValue = typeof tagDictionary[tag] !== 'undefined';
                    if (!hasTagDefaultValue) {
                        tagDictionary[tag] = '';
                    }
                }
                else {
                    for (var tagKey in tag) {
                        var tagValue = tagDictionary[tagKey];
                        var hasTagValue = tagValue && Object.keys(tagValue).length !== 0;
                        if (!hasTagValue) {
                            tagDictionary[tagKey] = tag[tagKey];
                        }
                    }
                }
            }
            return tagDictionary;
        };
        Registry.prototype._validateRegistration = function (key, type, registrationType) {
            var keyIsNotAString = typeof key !== 'string';
            if (keyIsNotAString) {
                throw new Error('Registration Key must be a string!');
            }
            var noTypeProvided = type === null || type === undefined;
            if (noTypeProvided) {
                throw new Error("The provided type for registration key '" + key + "' is undefined!");
            }
            switch (registrationType) {
                case expectedRegistrationType.class:
                    break;
                case expectedRegistrationType.factory:
                    if (typeof type !== 'function') {
                        throw new Error('Must pass a function to a factory registration!');
                    }
                    break;
                default:
                    break;
            }
        };
        return Registry;
    }());
    exports.Registry = Registry;
});
