// @process-engine/persistence_api.contracts/dist/commonjs/data_models/process_model/events/event_base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../base/index");
/**
 * The base class for all events.
 */
class Event extends index_1.FlowNode {
}
exports.Event = Event;
/**
 * The base class of all event definitions.
 */
class EventDefinition extends index_1.BaseElement {
}
exports.EventDefinition = EventDefinition;
/**
 * Contains the definition for an ErrorEvent.
 */
class ErrorEventDefinition extends EventDefinition {
}
exports.ErrorEventDefinition = ErrorEventDefinition;
/**
 * Contains the definition for a LinkEvent.
 */
class LinkEventDefinition extends EventDefinition {
    constructor(linkName) {
        super();
        this.name = linkName;
    }
}
exports.LinkEventDefinition = LinkEventDefinition;
/**
 * Contains the definition for a MessageEvent.
 */
class MessageEventDefinition extends EventDefinition {
}
exports.MessageEventDefinition = MessageEventDefinition;
/**
 * Contains the definition for a SignalEvent.
 */
class SignalEventDefinition extends EventDefinition {
}
exports.SignalEventDefinition = SignalEventDefinition;
/**
 * Contains the definition for a TerminateEndEvent.
 */
class TerminateEventDefinition extends EventDefinition {
}
exports.TerminateEventDefinition = TerminateEventDefinition;
/**
 * Contains the definition for a TimerEvent.
 */
class TimerEventDefinition extends EventDefinition {
    constructor() {
        super(...arguments);
        this.enabled = true;
    }
}
exports.TimerEventDefinition = TimerEventDefinition;
/**
 * Contains all timer types that are currently supported.
 */
var TimerType;
(function (TimerType) {
    TimerType[TimerType["timeCycle"] = 0] = "timeCycle";
    TimerType[TimerType["timeDate"] = 1] = "timeDate";
    TimerType[TimerType["timeDuration"] = 2] = "timeDuration";
})(TimerType = exports.TimerType || (exports.TimerType = {}));
