// @process-engine/management_api_http/dist/amd/endpoints/user_task/user_task_router.js
define(["require", "exports", "@essential-projects/http_node", "@process-engine/management_api_contracts", "async-middleware", "../../middlewares/resolve_identity"], function (require, exports, http_node_1, management_api_contracts_1, async_middleware_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class UserTaskRouter extends http_node_1.BaseRouter {
        constructor(userTaskController, identityService) {
            super();
            this.userTaskController = userTaskController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.userTaskController;
            this.router.get(management_api_contracts_1.restSettings.paths.processModelUserTasks, async_middleware_1.wrap(controller.getUserTasksForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.processInstanceUserTasks, async_middleware_1.wrap(controller.getUserTasksForProcessInstance.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.correlationUserTasks, async_middleware_1.wrap(controller.getUserTasksForCorrelation.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.processModelCorrelationUserTasks, async_middleware_1.wrap(controller.getUserTasksForProcessModelInCorrelation.bind(controller)));
            this.router.post(management_api_contracts_1.restSettings.paths.finishUserTask, async_middleware_1.wrap(controller.finishUserTask.bind(controller)));
        }
    }
    exports.UserTaskRouter = UserTaskRouter;
});
