// @essential-projects/errors_ts/dist/commonjs/error_serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_error_1 = require("./base_error");
const is_error_1 = require("./is_error");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function serializeError(error) {
    if (is_error_1.isEssentialProjectsError(error)) {
        return error.serialize();
    }
    if (error instanceof Error) {
        const errorJson = {
            message: error.message || '',
            name: error.name || '',
            stack: error.stack || '',
        };
        return JSON.stringify(errorJson);
    }
    return JSON.stringify(error);
}
exports.serializeError = serializeError;
function deserializeError(stringifiedError) {
    try {
        const result = base_error_1.BaseError.deserialize(stringifiedError);
        return result;
    }
    catch (error) {
        return deserializeBasicError(stringifiedError);
    }
}
exports.deserializeError = deserializeError;
function deserializeBasicError(stringifiedError) {
    const deserializedValue = tryParse(stringifiedError);
    const notAnError = !deserializedValue.name || !deserializedValue.message || !deserializedValue.stack;
    if (notAnError) {
        return deserializedValue;
    }
    const error = new Error(deserializedValue.message);
    error.stack = deserializedValue.stack;
    error.name = deserializedValue.name;
    return error;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function tryParse(value) {
    try {
        const result = JSON.parse(value);
        return result;
    }
    catch (error) {
        return value;
    }
}
