// @process-engine/consumer_api_http/dist/commonjs/endpoints/flow_node_instance/flow_node_instance_controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class FlowNodeInstanceController {
    constructor(flowNodenstanceService) {
        this.httpCodeSuccessfulResponse = 200;
        this.httpCodeSuccessfulNoContentResponse = 204;
        this.flowNodenstanceService = flowNodenstanceService;
    }
    async getAllSuspendedTasks(request, response) {
        const identity = request.identity;
        const offset = request.query.offset || 0;
        const limit = request.query.limit || 0;
        const result = await this.flowNodenstanceService.getAllSuspendedTasks(identity, offset, limit);
        response.status(this.httpCodeSuccessfulResponse).json(result);
    }
    async getSuspendedTasksForProcessModel(request, response) {
        const processModelId = request.params.process_model_id;
        const identity = request.identity;
        const offset = request.query.offset || 0;
        const limit = request.query.limit || 0;
        const result = await this.flowNodenstanceService.getSuspendedTasksForProcessModel(identity, processModelId, offset, limit);
        response.status(this.httpCodeSuccessfulResponse).json(result);
    }
    async getSuspendedTasksForProcessInstance(request, response) {
        const processInstanceId = request.params.process_instance_id;
        const identity = request.identity;
        const offset = request.query.offset || 0;
        const limit = request.query.limit || 0;
        const result = await this.flowNodenstanceService.getSuspendedTasksForProcessInstance(identity, processInstanceId, offset, limit);
        response.status(this.httpCodeSuccessfulResponse).json(result);
    }
    async getSuspendedTasksForCorrelation(request, response) {
        const correlationId = request.params.correlation_id;
        const identity = request.identity;
        const offset = request.query.offset || 0;
        const limit = request.query.limit || 0;
        const result = await this.flowNodenstanceService.getSuspendedTasksForCorrelation(identity, correlationId, offset, limit);
        response.status(this.httpCodeSuccessfulResponse).json(result);
    }
    async getSuspendedTasksForProcessModelInCorrelation(request, response) {
        const processModelId = request.params.process_model_id;
        const correlationId = request.params.correlation_id;
        const identity = request.identity;
        const offset = request.query.offset || 0;
        const limit = request.query.limit || 0;
        const result = await this.flowNodenstanceService.getSuspendedTasksForProcessModelInCorrelation(identity, processModelId, correlationId, offset, limit);
        response.status(this.httpCodeSuccessfulResponse).json(result);
    }
}
exports.FlowNodeInstanceController = FlowNodeInstanceController;
