// @process-engine/management_api_http/dist/amd/endpoints/cronjob/cronjob_router.js
define(["require", "exports", "async-middleware", "@essential-projects/http_node", "@process-engine/management_api_contracts", "../../middlewares/resolve_identity"], function (require, exports, async_middleware_1, http_node_1, management_api_contracts_1, resolve_identity_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class CronjobRouter extends http_node_1.BaseRouter {
        constructor(cronjobController, identityService) {
            super();
            this.cronjobController = cronjobController;
            this.identityService = identityService;
        }
        get baseRoute() {
            return 'api/management/v1';
        }
        async initializeRouter() {
            this.registerMiddlewares();
            this.registerRoutes();
        }
        registerMiddlewares() {
            const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
            this.router.use(async_middleware_1.wrap(resolveIdentity));
        }
        registerRoutes() {
            const controller = this.cronjobController;
            this.router.get(management_api_contracts_1.restSettings.paths.getActiveCronjobs, async_middleware_1.wrap(controller.getAllActiveCronjobs.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getCronjobExecutionHistoryForProcessModel, async_middleware_1.wrap(controller.getCronjobExecutionHistoryForProcessModel.bind(controller)));
            this.router.get(management_api_contracts_1.restSettings.paths.getCronjobExecutionHistoryForCrontab, async_middleware_1.wrap(controller.getCronjobExecutionHistoryForCrontab.bind(controller)));
        }
    }
    exports.CronjobRouter = CronjobRouter;
});
