// @process-engine/management_api_http/dist/commonjs/endpoints/event/event_router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const http_node_1 = require("@essential-projects/http_node");
const management_api_contracts_1 = require("@process-engine/management_api_contracts");
const async_middleware_1 = require("async-middleware");
const resolve_identity_1 = require("../../middlewares/resolve_identity");
class EventRouter extends http_node_1.BaseRouter {
    constructor(eventController, identityService) {
        super();
        this.eventController = eventController;
        this.identityService = identityService;
    }
    get baseRoute() {
        return 'api/management/v1';
    }
    async initializeRouter() {
        this.registerMiddlewares();
        this.registerRoutes();
    }
    registerMiddlewares() {
        const resolveIdentity = resolve_identity_1.createResolveIdentityMiddleware(this.identityService);
        this.router.use(async_middleware_1.wrap(resolveIdentity));
    }
    registerRoutes() {
        const controller = this.eventController;
        this.router.get(management_api_contracts_1.restSettings.paths.waitingProcessModelEvents, async_middleware_1.wrap(controller.getWaitingEventsForProcessModel.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.waitingCorrelationEvents, async_middleware_1.wrap(controller.getWaitingEventsForCorrelation.bind(controller)));
        this.router.get(management_api_contracts_1.restSettings.paths.waitingProcessModelCorrelationEvents, async_middleware_1.wrap(controller.getWaitingEventsForProcessModelInCorrelation.bind(controller)));
        this.router.post(management_api_contracts_1.restSettings.paths.triggerMessageEvent, async_middleware_1.wrap(controller.triggerMessageEvent.bind(controller)));
        this.router.post(management_api_contracts_1.restSettings.paths.triggerSignalEvent, async_middleware_1.wrap(controller.triggerSignalEvent.bind(controller)));
    }
}
exports.EventRouter = EventRouter;
