// @essential-projects/http_node/dist/amd/base_router.js
define(["require", "exports", "express"], function (require, exports, express) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class BaseRouter {
        constructor() {
            this.config = undefined;
            // eslint-disable-next-line @typescript-eslint/member-naming
            this._router = undefined;
        }
        get router() {
            if (!this._router) {
                // eslint-disable-next-line 6river/new-cap
                this._router = express.Router();
            }
            return this._router;
        }
        get baseRoute() {
            const baseRoute = this.config.baseRoute;
            if (!baseRoute) {
                return '';
            }
            return baseRoute;
        }
        initialize() {
            return this.invokeAsPromiseIfPossible(this.initializeRouter, this);
        }
        /**
         * Inheriting routers can override this method to cleanup on shutdown.
         */
        dispose() { }
        async invokeAsPromiseIfPossible(functionToInvoke, invocationContext, invocationParameter) {
            const isValidFunction = typeof functionToInvoke === 'function';
            if (!isValidFunction) {
                return Promise.resolve();
            }
            return await functionToInvoke.call(invocationContext, invocationParameter);
        }
    }
    exports.BaseRouter = BaseRouter;
});
