// @process-engine/management_api_contracts/dist/amd/data_models/flow_node_instance/bpmn_type.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var BpmnType;
    (function (BpmnType) {
        BpmnType["emptyActivity"] = "bpmn:Task";
        BpmnType["userTask"] = "bpmn:UserTask";
        BpmnType["exclusiveGateway"] = "bpmn:ExclusiveGateway";
        BpmnType["parallelGateway"] = "bpmn:ParallelGateway";
        BpmnType["serviceTask"] = "bpmn:ServiceTask";
        BpmnType["startEvent"] = "bpmn:StartEvent";
        BpmnType["endEvent"] = "bpmn:EndEvent";
        BpmnType["intermediateCatchEvent"] = "bpmn:IntermediateCatchEvent";
        BpmnType["intermediateThrowEvent"] = "bpmn:IntermediateThrowEvent";
        BpmnType["scriptTask"] = "bpmn:ScriptTask";
        BpmnType["boundaryEvent"] = "bpmn:BoundaryEvent";
        BpmnType["callActivity"] = "bpmn:CallActivity";
        BpmnType["subProcess"] = "bpmn:SubProcess";
        BpmnType["manualTask"] = "bpmn:ManualTask";
        BpmnType["complexGateway"] = "bpmn:ComplexGateway";
        BpmnType["inclusiveGateway"] = "bpmn:InclusiveGateway";
        BpmnType["eventBasedGateway"] = "bpmn:EventBasedGateway";
        BpmnType["sendTask"] = "bpmn:SendTask";
        BpmnType["receiveTask"] = "bpmn:ReceiveTask";
    })(BpmnType = exports.BpmnType || (exports.BpmnType = {}));
});
